README.md
# Alchemy Search

The search plugin for the Alchemy MVC

## Installation

Just installing the npm package can be done like this:

    $ npm install alchemy-search

## Activate

You can activate the plugin by adding this to the main `bootstrap.js` file:

```javascript
alchemy.usePlugin('search', {});
```

## Use

This plugin gives you a new behaviour `search` to add to your models,
you can do this after constituting your model.

You have to supply an options object when adding the behaviour.
Here you need to add the fields you want to make searchable and the importance of those fields.

```javascript
MyModel.constitute(function addFields() {

    // Add your fields like you normally would
    this.addField('name', 'String');
    this.addField('title', 'String', {translatable: true});
    this.addField('description', 'String', {translatable: true});

    // And finally add the search behaviour
    this.addBehaviour('search', {
        fields: {
            title       : 50,
            name        : 30,
            description : 10
        }
    });
});
```

Now, after a record of (in this example) `MyModel` has been saved, the search behaviour will create the required data in the database.

A new method has been added to the `Model` class: `touchAll(callback)`. It basically re-saves all the records in a model (without setting the `updated` date) and triggering the before- and after-save events.

Finally, you can search through the model like so:



CHANGELOG.md
## 0.2.0 (WIP)

* Add support for HTML fields
* Make compatible with Alchemy v1.1.0

## 0.1.0 (2018-08-27)

* Initial commit

model/document_list.js
/**
 * Add the search results of another list to this one
 *
 * @author   Jelle De Loecker <jelle@develry.be>
 * @since    0.1.0
 * @version  0.1.0
 *
 * @param    {Alchemy.DocumentList}   list   The other list to add
 *
 * @return   {Alchemy.DocumentList}   The new document list
 */
Classes.Alchemy.DocumentList.setMethod(function addSearchResultList(list) {

	var theirs,
	    match,
	    ours,
	    i,
	    j;

	for (i = 0; i < list.length; i++) {
		theirs = list[i];
		match = null;

		for (j = 0; j < this.length; j++) {
			ours = this[j];

			if (Object.alike(theirs._id, ours._id)) {
				match = ours;
				break;
			}
		}

		if (!match) {
			this.push(theirs);
			this.available++;
		} else {
			match.$record.__score += theirs.$record.__score;
		}
	}

	this.sortByPath(-1, '$record.__score');
});

/**
 * Get the best search result out of a document list
 *
 * @author   Jelle De Loecker <jelle@develry.be>
 * @since    0.1.0
 * @version  0.1.0
 *
 * @param    {Alchemy.DocumentList}   list   Optional other list
 *
 * @return   {Alchemy.Document}
 */
Classes.Alchemy.DocumentList.setMethod(function getBestSearchResult(list) {

	var context;

	if (list) {
		context = this.clone();
		context.addSearchResultList(list);
	} else {
		context = this;
	}

	return context[0];
});

/**
 * Get the best matching result out of this and another list
 *
 * @author   Jelle De Loecker <jelle@develry.be>
 * @since    0.1.0
 * @version  0.1.0
 *
 * @param    {Alchemy.DocumentList}   list   Optional other list
 *
 * @return   {Alchemy.Document}
 */
Classes.Alchemy.DocumentList.setMethod(function getBestSearchResultFromBothLists(list) {

	var results = [],
	    theirs,
	    match,
	    clone,
	    ours,
	    i,
	    j;

	clone = this.clone();

	for (i = 0; i < list.length; i++) {
		theirs = list[i];
		match = null;

		for (j = 0; j < clone.length; j++) {
			ours = clone[j];

			// @TODO: this should work with Object.alike
			if (String(theirs._id) == String(ours._id)) {
				match = ours;
				break;
			}
		}

		// We're only interested if it's a result in both!
		if (match) {
			match.$record.__score += theirs.$record.__score;
			results.push(match);
		}
	}

	results.sortByPath(-1, '$record.__score');

	return results[0];
});

behaviour/search_behaviour.js
var getMetaphone = alchemy.use('double-metaphone'),
    all_prefixes = alchemy.shared('Routing.prefixes');

/**
 * The Search Behaviour class
 *
 * @constructor
 * @extends       Alchemy.Behaviour
 *
 * @author        Jelle De Loecker   <jelle@develry.be>
 * @since         0.1.0
 * @version       0.1.0
 */
var Search = Function.inherits('Alchemy.Behaviour', function SearchBehaviour(model, options) {
	Behaviour.call(this, model, options);
});

/**
 * Called when this behaviour is attached to a schema
 *
 * @author   Jelle De Loecker   <jelle@develry.be>
 * @since    0.1.0
 * @version  0.1.0
 */
Search.setStatic(function attached(schema, options) {

	if (!schema.model_class) {
		return;
	}

	schema.model_class.setMethod(function search(config, options, callback) {

		var search_results,
		    datasource,
		    collection,
		    thisModel = this,
		    pointers = {},
		    Search = this.getModel('AlSearch'),
		    query,
		    ids;

		if (typeof config == 'string') {
			query = config;
			config = {};
		} else {
			query = config.query;
		}

		if (typeof options == 'function') {
			callback = options;
			options = {};
		}

		query = query.trim();

		datasource = thisModel.datasource;

		Function.series(function getCollection(next) {
			datasource.collection(Search.table, function gotCollection(err, _collection) {

				if (err) {
					return next(err);
				}

				collection = _collection;
				next();
			});
		}, function doAggregate(next) {

			var rx_keywords,
			    mp_matchers,
			    keywords,
			    pipeline,
			    matchers,
			    rx_piece,
			    pieces,
			    piece,
			    i;

			pieces = query.split(/[\s+,]/).filter(Boolean);
			keywords = [];
			rx_keywords = [];
			mp_matchers = [];
			matchers = [];

			matchers.push({
				name  : 'exact',
				field : 'value',
				value : query,
				mod   : 100
			});

			if (query.length > 2) {
				matchers.push({
					name  : 'starts_with',
					field : 'value',
					rx    : RegExp.interpret('^' + query, 'i'),
					start : true,
					value : query,
					mod   : 40
				});
			}

			if (query.length > 3) {
				matchers.push({
					name  : 'contains',
					field : 'value',
					rx    : RegExp.interpret(query, 'i'),
					value : query,
					mod   : 30
				});
			}

			// Do regular pieces & rx pieces
			for (i = 0; i < pieces.length; i++) {
				piece = pieces[i];

				if (piece.length > 2) {
					rx_piece = RegExp.interpret(piece, 'i');
					keywords.push(piece);
					rx_keywords.push(rx_piece);

					matchers.push({
						name  : 'starts_with_piece',
						field : 'value',
						rx    : RegExp.interpret('^' + piece, 'i'),
						start : true,
						value : piece,
						mod   : Math.min(15, piece.length * 8)
					});

					matchers.push({
						name  : 'identical_piece',
						field : 'value',
						value : piece,
						mod   : Math.min(40, piece.length * 8)
					});

					matchers.push({
						name  : 'rx_piece',
						field : 'value',
						rx    : rx_piece,
						value : piece,
						mod   : Math.min(20, piece.length * 8)
					});
				}
			}

			// Now do metaphone pieces
			for (i = 0; i < pieces.length; i++) {
				piece = pieces[i];

				// Only use larger words for metaphone search
				if (piece.length < 4) {
					continue;
				}

				piece = getMetaphone(pieces[i]).join(' ').trim();

				matchers.push({
					name  : 'identical_mp',
					field : 'metaphone',
					value : piece,
					meta  : true,
					mod   : 15
				});

				matchers.push({
					name  : 'starts_with_mp',
					field : 'metaphone',
					rx    : RegExp.interpret('^' + piece, 'i'),
					start : true,
					value : piece,
					meta  : true,
					mod   : 8
				});

				matchers.push({
					name  : 'contains_mp',
					field : 'metaphone',
					rx    : RegExp.interpret(piece, 'i'),
					value : piece,
					meta  : true,
					mod   : 2
				});
			}

			// Sort the matchers, the highest modder should be checked first
			matchers.sortByPath(-1, 'mod');

			let mp_branches = [],
			    branches = [],
			    $project,
			    matcher,
			    index,
			    temp,
			    $or = [];

			for (i = 0; i < matchers.length; i++) {
				matcher = matchers[i];

				// Add the strict equal match
				temp = {};
				temp[matcher.field] = matcher.value;
				$or.push(temp);

				let value = matcher.value;
				let field = '$' + matcher.field;

				if (!matcher.meta) {
					value = value.toLowerCase();
					field = {$toLower: field};
				}

				if (matcher.rx != null) {
					// Add the regex to the matchers
					temp = {};
					temp[matcher.field] = matcher.rx;
					$or.push(temp);

					// Add it to the switch cashe using indexOf
					if (matcher.start) {
						temp = {
							$eq : [
								{$indexOfCP: [field, value]},
								0
							]
						};
					} else {
						temp = {
							$gt : [
								{$indexOfCP: [field, value]},
								-1
							]
						};
					}
				} else {
					temp = {
						$eq: [field, value]
					};
				}

				if (matcher.meta) {
					mp_branches.push({
						case : temp,
						then : matcher.mod
					});
				} else {
					branches.push({
						case : temp,
						then : matcher.mod
					});
				}
			}

			$project = {
				field_name : 1,
				record_id  : 1,
				value      : 1,
				weight     : 1
			};

			if (branches.length) {
				$project.mod = {
					$switch : {
						branches : branches,
						default  : 0
					}
				};
			}

			if (mp_branches.length) {
				$project.mp_mod     = {
					$switch : {
						branches : mp_branches,
						default  : 0
					}
				};
			}

			let $match = {
				model : thisModel.name,
				$or   : $or
			};

			// Limit the search to certain fields if needed
			if (options.search_fields) {
				let fields = Array.cast(options.search_fields);
				$match.field_name = {$in: fields};
			}

			pipeline = [
				{
					$match : $match
				},
				{
					$project : $project
				},
				{
					$project : {
						field_name : 1,
						record_id  : 1,
						value      : 1,
						weight     : 1,
						mod : 1,
						mp_mod: 1,
						score : {
							$add : [
								{$multiply: ['$mod', '$weight']},
								{$multiply: ['$mp_mod', '$weight']}
							]
						}
					}
				},
				{
					$group : {
						_id   : "$record_id",
						score : {$sum : '$score'}
					}
				},
				{
					$sort : {
						'score' : -1
					}
				}
			];

			if (config.skip) {
				pipeline.push({$skip: config.skip});
			}

			if (config.limit) {
				pipeline.push({$limit: config.limit});
			}

			collection.aggregate(pipeline, function gotResults(err, cursor) {

				if (err) {
					return next(err);
				}

				cursor.toArray(function gotArray(err, results) {

					if (err) {
						return next(err);
					}

					search_results = results;
					ids = [];

					for (i = 0; i < results.length; i++) {
						ids.push(results[i]._id);
					}

					next();
				});
			});
		}, function done(err) {

			if (err) {
				return callback(err);
			}

			let final_options = JSON.clone(options);
			let want_document = final_options.document;

			if (want_document !== false) {
				want_document = true;
			}

			if (!final_options.conditions) {
				final_options.conditions = {};
			}

			// Always overwrite the _ids to look for
			final_options.conditions._id = ids;

			final_options.document = want_document;

			thisModel.find('all', final_options, function gotFinalResults(err, results) {

				if (err) {
					return callback(err);
				}

				let record,
				    temp,
				    i,
				    j;

				for (i = 0; i < results.length; i++) {
					record = results[i];

					for (j = 0; j < search_results.length; j++) {
						temp = search_results[j];

						if (String(temp._id) == String(record._id)) {
							record.$record.__score = temp.score;
							break;
						}
					}
				}

				results.sortByPath(-1, '$record.__score');

				callback(null, results);
			});
		});
	});
});

/**
 * Called after the model saves a record.
 *
 * @author   Jelle De Loecker   <jelle@develry.be>
 * @since    0.1.0
 * @version  0.1.1
 */
Search.setMethod(function afterSave(record, options, created) {

	var that = this,
	    next = this.wait();

	// Find the complete saved item
	Model.get(that.model.name).findById(record._id, function gotFullRecord(err, result) {

		if (err) {
			return next(err);
		}

		if (!result) {
			return next();
		}

		result = result[that.model.name];

		let Search = Model.get('AlSearch');

		Function.series(function deleteOld(next) {
			Search.find('all', {conditions: {record_id: record._id}}, function gotOldRecords(err, records) {

				if (err) {
					return next(err);
				}

				records.forEach(function eachRecord(record) {
					record.remove();
				});

				next();
			});
		}, function createNewRecords(next) {

			for (let field_name in that.options.fields) {
				let field = that.model.getField(field_name);

				if (!field) {
					return next(new Error('Field "' + field_name + '" does not exist, can not add to search'));
				}

				let weight = that.options.fields[field_name];
				let value = result[field_name];

				if (value == '' || value == null) {
					continue;
				}

				let data = {
					model      : that.model.name,
					record_id  : record._id,
					field_name : field_name,
					weight     : weight
				};

				let is_html = field.constructor.name == 'HtmlFieldType';

				if (field.is_translatable) {

					// Should be an object. If it's not, skip it
					if (!Object.isPlainObject(value)) {
						continue;
					}

					for (let prefix in value) {
						let prefix_value = value[prefix];

						if (prefix_value == '' || prefix_value == null) {
							continue;
						}

						if (is_html && typeof prefix_value == 'string') {
							prefix_value = prefix_value.stripTags(false, ' ').decodeHTML();
						}

						saveValue(prefix_value, data, prefix);
					}
				} else {

					if (is_html && value && typeof value == 'string') {
						value = value.stripTags(false, ' ').decodeHTML();
					}

					saveValue(value, data);
				}
			}

			next();
		}, next)

		function saveValue(value, data, language) {

			data = Object.assign({}, data);

			data.value = value;

			if (typeof value == 'string') {
				let pieces = value.split(/[\s+,]/).filter(Boolean);
				let metaphones = [];
				let piece;

				for (let i = 0; i < pieces.length; i++) {
					piece = pieces[i];

					if (piece.length < 4) {
						continue;
					}

					metaphones.push(getMetaphone(pieces[i]).join(' '));
				}

				metaphones = metaphones.join(' ').trim();

				if (metaphones) {
					data.metaphone = metaphones;
				}
			}

			if (language) {
				data.language = language;
			}

			Search.save(data);
		}
	});
});

model/al_search_model.js
/**
 * The Alchemy Search Model class
 *
 * @constructor
 *
 * @author   Jelle De Loecker   <jelle@develry.be>
 * @since    0.1.0
 * @version  0.1.0
 */
var Search = Function.inherits('Alchemy.Model.App', function AlSearch(conduit, options) {
	AlSearch.super.call(this, conduit, options);
});

/**
 * Constitute the class wide schema
 *
 * @author   Jelle De Loecker <jelle@develry.be>
 * @since    0.1.0
 * @version  0.1.0
 */
Search.constitute(function addFields() {

	// The model origin
	this.addField('model', 'String');

	// The original record id
	this.addField('record_id', 'ObjectId');

	// The field name
	this.addField('field_name', 'String');

	// The weight of this entry
	this.addField('weight', 'Number');

	// The field value
	this.addField('value', 'String');

	// The metaphone value
	this.addField('metaphone', 'String');

	// Optional language
	this.addField('language', 'String');

	// Add indexes
	this.addIndex('model');
	this.addIndex('record_id');
	this.addIndex('field_name');
});

/**
 * Do a simple search
 *
 * @author   Jelle De Loecker <jelle@develry.be>
 * @since    0.1.0
 * @version  0.1.0
 */
Search.setMethod(function simpleSearch(config, callback) {

	var that = this,
	    default_conditions,
	    rx_keywords,
	    key_name,
	    keywords,
	    results,
	    fields,
	    pieces,
	    tasks,
	    piece,
	    query,
	    i;

	results = {};
	default_conditions = config.conditions || {};
	query = config.query.trim();
	fields = Array.cast(config.fields);
	pieces = query.split(/[\s+,]/).filter(Boolean);
	rx_keywords = [];
	keywords = [];
	tasks = [];
	key_name = config.key || '_id';

	for (i = 0; i < pieces.length; i++) {
		piece = pieces[i];

		if (piece.length > 2) {
			keywords.push(piece);
			rx_keywords.push(RegExp.interpret(piece, 'i'));
		}
	}

	for (let i = 0; i < fields.length; i++) {
		let field_name = fields[i];

		tasks.push(function doExact(next) {

			var conditions = Object.assign({}, default_conditions);
			conditions[field_name] = query;

			// Look for the entire query
			findRecords(100, conditions, next);
		});

		tasks.push(function doNearlyExact(next) {

			var conditions = Object.assign({}, default_conditions);
			conditions[field_name] = RegExp.interpret(query, 'i');

			findRecords(70, conditions, next);
		});

		tasks.push(function startsWith(next) {

			var conditions = Object.assign({}, default_conditions);
			conditions[field_name] = RegExp.interpret('^' + query, 'i');

			findRecords(15, conditions, next);
		});

		// Don't do individual keywords if there is only 1 keyword
		if (keywords.length == 1) {
			continue;
		}

		for (let i = 0; i < keywords.length; i++) {
			let keyword = keywords[i];
			let rx_keyword = rx_keywords[i];

			tasks.push(function startsWith(next) {

				var conditions = Object.assign({}, default_conditions);
				conditions[field_name] = RegExp.interpret('^' + keyword, 'i');

				findRecords(15, conditions, next);
			});

			// Search for identical matches
			tasks.push(function doKeyword(next) {
				var conditions = Object.assign({}, default_conditions);
				conditions[field_name] = keyword;

				findRecords(40, conditions, next);
			});

			// Search for general matches
			tasks.push(function doGeneralKeyword(next) {
				var conditions = Object.assign({}, default_conditions);
				conditions[field_name] = rx_keyword;

				findRecords(20, conditions, next);
			});
		}
	}

	Function.parallel(4, tasks, function done(err) {

		if (err) {
			return callback(err);
		}

		let record,
		    key;

		results = Object.values(results);
		results.sortByPath(-1, '__score');

		callback(null, results);
	});

	// Find with the given conditions
	function findRecords(importance, conditions, options, next) {

		var def_options = {
			recursive : 0,
			document  : false,
			limit     : 50,
			available : false
		};

		if (typeof options == 'function') {
			next = options;
			options = {};
		}

		options = Object.assign({}, def_options, options);
		options.conditions = conditions;

		if (config.query_fields) {
			options.fields = config.query_fields;
		}

		that.find('all', options, function gotResult(err, records) {
			scoreResults(err, records, importance, next);
		});
	}

	// Score the found records
	function scoreResults(err, records, importance, next) {

		var entry,
		    i;

		if (err) {
			return next(err);
		}

		for (i = 0; i < records.length; i++) {
			entry = records[i][that.name];

			if (!results[entry[key_name]]) {
				results[entry[key_name]] = entry;
				entry.__hits = 0;
				entry.__score = 0;
			} else {
				entry = results[entry[key_name]];
			}

			entry.__hits++;
			entry.__score += importance;
		}

		next();
	}

	function getSimilarCharPercentage(str_a, str_b) {

		var total_length = str_a.length + str_b.length,
		    result;

		result = (_countSimilarChars(str_a, str_b) + _countSimilarChars(str_b, str_a));

		result = (result / total_length) * 100;

		return result;
	}

	function _countSimilarChars(str_a, str_b) {

		var index,
		    result = 0,
		    a,
		    b,
		    i;

		a = str_a.toLowerCase().split('');
		b = str_b.toLowerCase().split('');

		for (i = 0; i < a.length; i++) {
			index = b.indexOf(a[i]);

			if (index > -1) {
				result++;
				b.splice(index, 1);
			}
		}

		return result;
	}
});

/**
 * Re-save all records in the database
 * without updating the 'updated' field
 *
 * @author   Jelle De Loecker <jelle@develry.be>
 * @since    0.1.0
 * @version  0.1.0
 */
Model.setMethod(function touchAll(callback) {

	var i = 1;

	if (!callback) {
		callback = Function.thrower;
	}

	this.find('all', {limit: 200}, function gotRecords(err, result) {

		var tasks = [];

		if (!result.length) {
			return callback(null);
		}

		result.forEach(function eachRecord(record) {
			tasks.push(function doSave(next) {
				record.save(null, {set_updated: false}, next);
			});
		});

		Function.parallel(8, tasks, function done(err) {

			if (err) {
				return callback(err);
			}

			result.findNextBatch(gotRecords);
		});
	});
});
